Invenio3 formatter with angular

ioannistsanaktsidis opened the pull request.

kaplun commented on inspirehep/modules/records/serializers/schemas/latexus_schema.py:
```diff
@@ -0,0 +1,35 @@
+class RecordSchemaLATEXUSV1(Schema):
+    """Schema for latexus v1 in JSON."""
+
+    latexus = fields.Str()
```
Is this a stub?

ioannistsanaktsidis replied:
Nope, it is not. This  """Marshmallow bibtex schema."""  should be  """Marshmallow latexus schema."""

jmartinm commented on inspirehep/modules/formatter/views.py:
```diff
@@ -0,0 +1,125 @@
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+import itertools
```
All of these will disappear since we use the `invenio-records-rest` API right?

By passing the `Accept` header with the format we want it will return bibtex, latex and so on.

ioannistsanaktsidis replied:
Yes exactly, all of them will disappear. I just keep them as reference now to check the logic sometimes. I will remove them when it is ready for merging.

jmartinm replied:
And if it is not needed we can probably remove the whole `formatter` module.

ioannistsanaktsidis replied:
Could be the case, everything now works with the serializers. Formatter will not be needed anymore in the end of the day.

jmartinm commented on inspirehep/modules/records/serializers/bibtex_serializer.py:
```diff
@@ -0,0 +1,43 @@
+        :param links_factory: Factory function for the link generation,
+                              which are added to the response.
+        """
+        return Bibtex(record).format()
```
Did you check that this works? As the schema has the field `bibtex` - https://github.com/inspirehep/inspire-next/pull/1024/files#diff-cc384d6e061b5ef5cce6d004dfa7984aR35 I am wondering if you need to return a structure with such a field inside. (like we do for brief response https://github.com/inspirehep/inspire-next/blob/invenio3/inspirehep/modules/records/serializers/json.py#L79

ioannistsanaktsidis replied:
I don't think it is necessary. I tried it like this and seems to work good.

jmartinm commented on inspirehep/utils/bibtex.py:
```diff
@@ -0,0 +1,679 @@
+                try:
+                    if journal and (volume != '' or pages != ''):
+                        recid = self.record['control_number', '']
+                        record = get_es_record_from_doc_type('literature',
```
How does this work? Can it be `literature` and `journals` ?

ioannistsanaktsidis replied:
Well when I ported it for invenio3 I had it like this:
```
if journal and (volume != '' or pages != ''):
    recid = self.record['control_number', '']
    pid = PersistentIdentifier.get(
                            'literature', str(recid))
                        record = es.get_source(
                            index='records-journals',
                            id=str(pid.object_uuid),
                            doc_type='journals')
```
I introduced this 
```
def get_es_record_from_doc_type(record_type, recid, collection):
```
so as to do the same job.
Makes sense?

jmartinm replied:
I am not sure in that example. Maybe it would make more sense to me if the code was:

```python
pid = PersistentIdentifier.get(
                            'journals', str(recid))
```

So that everything is related to `journals`. 

Otherwise I am not sure how the document will be found in the ES index.

ioannistsanaktsidis replied:
Ok, will check it again then when I apply @jacquerie  tests.

jmartinm commented on inspirehep/utils/record_getter.py:
```diff
@@ -73,6 +73,17 @@ def get_es_record(record_type, recid):
 
 
 @raise_record_getter_error_and_log
+def get_es_record_from_doc_type(record_type, recid, collection):
```
What is the difference between this one and `get_es_record` ?

jmartinm replied:
ping @ioannistsanaktsidis ;)

ioannistsanaktsidis replied:
I introduced this one so as to take care of cases like this :
```
 recid = self.record['control_number', '']
    pid = PersistentIdentifier.get(
                            'literature', str(recid))
                        record = es.get_source(
                            index='records-journals',
                            id=str(pid.object_uuid),
                            doc_type='journals')
```
Makes sense?

jmartinm commented:
There is a bit of a mixture of changes in CSS, JS, serializers and so on. Can you rebase and clean it up to just take care of the serializers?

ioannistsanaktsidis commented:
This PR should be merged together with this one https://github.com/inspirehep/inspire-next/pull/1003

ioannistsanaktsidis commented:
So this PR is failing because of the clipboard dependency. We have to merge this one https://github.com/inspirehep/inspirehep-js/pull/3 and I will push force again. After that the PR should not be failing.

kaplun commented on inspirehep/utils/bibtex.py:
```diff
@@ -0,0 +1,679 @@
+                except:
+                    return ''
+        else:
+            return ''
```
@ioannistsanaktsidis there is a problem with this function: currently it will return only the `coden` when available, otherwise it will only return `"". As it is currently written will never return the pubnote.

ioannistsanaktsidis replied:
Thats why I needed the knowledge module, so as to use this https://github.com/inspirehep/inspire/blob/master/bibformat/format_elements/bfe_INSPIRE_publi_coden.py#L57

kaplun replied:
OK, but what I mean is that this function is not returning a pubnote at all :smile: should it be rather called `_get_coden`?

kaplun replied:
Ah no sorry. Now I see: I was misleaded by:
```python
coden = ','.join(['coden'][0], volume, pages])
```

that variable should actually be called pubnote (since it's the union of a journal title/coden, a volume and pages)

ioannistsanaktsidis replied:
Cool :+1:

ioannistsanaktsidis commented:
@jmartinm  this one is ready, I just need to wait the tests from @jacquerie  and then we can port it.

jmartinm commented on inspirehep/modules/search/static/js/search/app.js:
```diff
@@ -22,9 +22,6 @@
  */
 
 require([
-    'angular',
```
are you sure about removing this?

ioannistsanaktsidis replied:
I can't see anything strange happening when I remove them...Do you think that they should be there so as to be initialized when the app is created ?

jmartinm replied:
Should be fine as the dependencies are defined in `settings.js` so I think they can be removed.

jmartinm commented on inspirehep/modules/theme/bundles.py:
```diff
@@ -63,7 +63,8 @@
         "flightjs": "~1.5.0",
-        "angular": "~1.4.8",
+        "angular": "~1.4.9",
+        "angular-sanitize": "~1.4.9",
```
In principle I don't think we need to add this. `inspirehep-search-js` already depends on that package so it gets installed.

jmartinm commented on inspirehep/modules/theme/jinja2filters.py:
```diff
@@ -45,6 +45,10 @@
 from .views import blueprint
 from inspirehep.utils.template import render_macro_from_template
 from inspirehep.modules.records.json_ref_loader import replace_refs
+from inspirehep.utils.bibtex import Bibtex
```
can you keep the order of imports? (place these under `inspirehep.utils`)

jmartinm commented:
Looks good overall, have not checked in detail the actual export functions as they are the same we had in invenio 2.x and waiting for @jacquerie 's tests

jacquerie commented:
Question: the new approach using serializers means we can get rid of the Jinja2 filters? https://github.com/inspirehep/inspire-next/blob/df77c34f7062553c9f3d1041db35a20cdc76730c/inspirehep/modules/theme/jinja2filters.py#L217-L234

ioannistsanaktsidis commented:
I guess yeah, there is no need of them. I will remove them from the PR.

jacquerie commented:
Good, four less lines of untested code :dancer:

ioannistsanaktsidis commented:
Cool :+1:

jacquerie commented on inspirehep/utils/export.py:
```diff
@@ -0,0 +1,149 @@
+    required field doesn't exist in the record.
+    """
+
+    def _init_(self, field):
```
Double underscores here...

jacquerie commented on inspirehep/utils/export.py:
```diff
@@ -0,0 +1,149 @@
+    def _init_(self, field):
+        self.field = field
+
+    def _str_(self):
```
...and here.

kaplun commented:
> Good, four less lines of untested code :dancer:

> ... Coverage decreased (-8.4%) to 48.914% when 

Mmh :confused:

jacquerie commented:
> Mmh :confused:

?

> Coverage decreased (-0.5%) to 56.861%

(and, by the end of tomorrow, that number will be positive)

kaplun commented:
Was just joking on: less lines to test, and the next line from coveralls was about the fact that coverage decreased :dancer:

jacquerie commented:
I don't think this is WIP anymore :tada:

This implements a core feature of INSPIRE, so I think we should merge this as soon as possible. Truth to be told, I'm not satisfied by the architecture of some parts of the code, but @ioannistsanaktsidis has not a lot of time left with us to fix those. The extensive unit tests should help us refactor anyway.

BTW, a few bugs I found are recorded as `xfailed` tests. We might want to fix at least those before AB.

ioannistsanaktsidis commented:
I need to add another function to the serializers for importing multiple
formats in one go. Will add it on Monday and then it can go on ;)
On Apr 15, 2016 5:54 PM, "Jacopo Notarstefano" <notifications@github.com>
wrote:

> I don't think this is WIP anymore [image: :tada:]
>
> This implements a core feature of INSPIRE, so I think we should merge this
> as soon as possible. Truth to be told, I'm not satisfied by the
> architecture of some parts of the code, but @ioannistsanaktsidis
> <https://github.com/ioannistsanaktsidis> has not a lot of time left with
> us to fix those. The extensive unit tests should help us refactor anyway.
>
> BTW, a few bugs I found are recorded as xfailed tests. We might want to
> fix at least those before AB.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/1024#issuecomment-210518117>
>

ioannistsanaktsidis commented:
So :) I added the functionality for exporting multiple formats at once. This one will reduce our JS code by a lot as some staff that I implemented in Angular are not needed anymore. Cheers 💃

jmartinm commented on inspirehep/modules/records/serializers/bibtex_serializer.py:
```diff
@@ -0,0 +1,55 @@
+        for hit in search_result['hits']['hits']:
+            records.append(Bibtex(record=hit['_source']).format())
+
+        return "\n".join(records)
```
Isn't it better if from the API it comes back as a list and then on the UI we iterate and display them?
